source/pkg/adapter/adapter.go

```diff
@@ -129,66 +121,93 @@ func (a *Adapter) Start(ctx context.Context) error {
 			a.logger.Info("Listening for messages", zap.String("consumerName", consumerName))
 
 			for {
-				//TODO: Check if statefulset's adapter pod is killed, and do not read more messages (to finish up work)
 				select {
-				case <-shutdownChannel:
-					a.logger.Info("Shutdown consumer", zap.String("consumerName", consumerName))
+				case <-ctx.Done(): //received a SIGINT or SIGTERM signal. Need to process pending messages and shut down consumer group
+
+					morePendingMsgs := true
+					for morePendingMsgs {
+						reply, err := conn.Do("XREADGROUP", "GROUP", groupName, consumerName, "COUNT", count, "BLOCK", blockms, "STREAMS", streamName, xreadID)
+						if err != nil {
+							a.logger.Error("Cannot read from stream", zap.Error(err))
+							continue
+						}
+
+						event, err := a.toEvent(reply)
+						if err != nil {
+							if strings.Contains(strings.ToLower(err.Error()), "number of items not equal to one (got 0)") || // no more pending messages!
+								strings.Contains(strings.ToLower(err.Error()), "expected a reply of type array") { // Xreadgroup timed out blocking after 5 seconds
+								morePendingMsgs = false
+							} else {
+								a.logger.Error("Cannot convert reply", zap.Error(err))
+							}
+							continue
+						}
+
+						if result := a.client.Send(ctx, *event); !cloudevents.IsACK(result) { //  Event is lost
+							a.logger.Error("Failed to send cloudevent", zap.Any("result", result))
+						}
+
+						_, err = conn.Do("XACK", streamName, groupName, event.ID())
+						if err != nil {
+							a.logger.Error("Cannot ack message", zap.Error(err))
+							morePendingMsgs = true
+							continue
+						}
+						a.logger.Info("Consumer acknowledged the message", zap.String("consumerName", consumerName))
```

Comment:
This code looks very similar to the code below. Just checking `xreadID == ">"` should be enough to exit the goroutine. WDYT?

Reply:
Yep thats enough :)